internal/backend/config/set.go

```diff
@@ -0,0 +1,77 @@
+package config
+
+import (
+	"fmt"
+
+	"go.autokitteh.dev/autokitteh/sdk/sdkerrors"
+)
+
+type ComponentConfig interface {
+	Validate() error
+}
+
+type emptyComponentConfig struct{}
+
+func (emptyComponentConfig) Validate() error { return nil }
+
+var EmptyComponentConfig emptyComponentConfig
+
+type baseSet interface {
+	hasDefault() bool
+	default_() ComponentConfig
+
+	hasDev() bool
+	dev() ComponentConfig
+
+	hasTest() bool
+	test() ComponentConfig
+}
+
+type Set[T ComponentConfig] struct {
+	Default, Dev, Test *T
+}
+
+func (s Set[T]) default_() ComponentConfig { return *s.Default }
+func (s Set[T]) dev() ComponentConfig      { return *s.Dev }
+func (s Set[T]) test() ComponentConfig     { return *s.Test }
+
+func (s Set[T]) hasDefault() bool { return s.Default != nil }
+func (s Set[T]) hasDev() bool     { return s.Dev != nil }
+func (s Set[T]) hasTest() bool    { return s.Test != nil }
+
+var EmptySet = Set[emptyComponentConfig]{Default: &EmptyComponentConfig}
+
+func (set *Set[T]) Choose(mode Mode) (chosen T, err error) {
```

Comment:
Why do we need this and not access `s.Default` etc?

Reply:
it's choosing the right one depending on mode - not sure i understand the q.